### PR TITLE
Add option for nightly, ci and dev test

### PR DESCRIPTION
**Pull request by yogeshsajanikar**
# Description

As a developer, one needs to run a fast quickcheck test to check if a commit is correct. For a CI, this number of quickcheck tests can be more. A nightly test can even be more aggressive in terms of the number of tests to run. To facilitate running different number of tests depending upon the environment, this PR adds an option "--test-env". A "--test-env" option can be set to either "nightly", "ci" or "dev". The "dev" environment is a default value. 

If "nightly" environment is set, then the number of quickcheck tests to run is set to "100_000". For "ci" environment this value is "10_000". The default "dev" environment sets the number of quickcheck tests to "_100_". An individual test may customise the number of quick check tests proportional to the time taken by the test to run, and to suit the environment.

The goal of current PR is to provide an option to set test environment and not to _tune_ the number of tests for different environment. "Tuning" the tests can be done as a separate task.

# Checklist

- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [x] New tests are added if needed and existing tests are updated
    - [x] If this branch changes Consensus and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../ouroboros-consensus/docs/interface-CHANGELOG.md)
    - [x] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [x] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [x] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [x] Reviewer requested

### CODE REVIEW

**Comment by yogeshsajanikar** [comment]
Things to be done
- Fix the number of tests for CI/Nightly
- Time the tests with CI/Nightly tests 
- Few tests run for a long time, nightly tests adjust quick check tests assuming it to be 100. This will be done by running tests.

File: ouroboros-consensus-test/src/Test/Util/TestMode.hs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import           Test.Tasty.Ingredients
+import           Test.Tasty.Options
+import           Test.Tasty.QuickCheck
+import           Test.Util.Nightly
+
+-- | 'defaultMain' extended with 'iohkTestEnvIngredient'
+defaultMainWithIohkTestMode :: TestTree -> IO ()
+defaultMainWithIohkTestMode testTree =
+    defaultMainWithIngredients (iohkNightlyIngredient : iohkTestModeIngredient : defaultIngredients) $
+    askIohkTestMode (defaultTestEnv $ askOption iohkNightlyToTestMode)
```
Comment (yogeshsajanikar) [inline ouroboros-consensus-test/src/Test/Util/TestMode.hs:23]: This doesn't work. The runtime option always returns `dev`, when `iohk-enable-nightly-tests` is set.

File: ouroboros-consensus-test/src/Test/Util/Nightly.hs [hunk 2]
```diff
@@ -35,7 +35,8 @@ instance IsOption IohkNightlyEnabled where
   defaultValue = IohkNightlyEnabled False
   parseValue = fmap IohkNightlyEnabled . safeReadBool
   optionName = pure "iohk-enable-nightly-tests"
```
Comment (dnadales) [inline ouroboros-consensus-test/src/Test/Util/Nightly.hs:37]: We should consider removing this option if it is not used anywhere.

Comment (dnadales) [reply ouroboros-consensus-test/src/Test/Util/Nightly.hs:37]: I don't see this module being used at all in our code.

File: ouroboros-consensus-test/src/Test/Util/TestMode.hs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  CI      -> adjustOption (const (QuickCheckTests 10000)) testTree
+  Dev     -> testTree
+
+-- | Reset quickcheck tests
+resetQuickCheckTests :: (QuickCheckTests -> QuickCheckTests) -> TestTree -> TestTree
+resetQuickCheckTests f = adjustOption $ const (f (QuickCheckTests 100))
+
+-- | An 'Option' that indicates the environment in which to run tests.
+data IohkTestMode =
+  -- | Run more expensive tests
```
Comment (dnadales) [inline ouroboros-consensus-test/src/Test/Util/TestMode.hs:56]: Maybe we can skip these hadocks. Isn't their meaning obvious from the context?

**Review by dnadales — COMMENTED** [review]
Thanks! Nice work. Some comments we should address:

-   Check we're respecting the style guide regarding `where` clauses and new data type defs.
-   Consider defining and using a function like [this one](https://github.com/input-output-hk/ouroboros-network/commit/fb56c188a83c1aa11bd5e240882cb4c9f7c961b0).
-   Remove Iohk from the function and module names.
-   Review haddocks.

Regarding the TODO's:

-   Fix the number of tests for CI/Nightly: we could consider the goal of this issue to remove the use of `withMaxSuccess` in favour of the tasty options.
-   However, I would not mind combing through the tests, timing them, elaborating a table of how much time each test takes, and proposing the adjustments for CI, Dev, and Nightly.

File: ouroboros-consensus-byron-test/test/Main.hs [hunk 4]
```diff
@@ -6,9 +6,10 @@ import qualified Test.Consensus.Byron.Golden (tests)
 import qualified Test.Consensus.Byron.Serialisation (tests)
 import qualified Test.ThreadNet.Byron (tests)
 import qualified Test.ThreadNet.DualByron (tests)
+import           Test.Util.TestMode (defaultMainWithIohkTestMode)
 
 main :: IO ()
-main = defaultMain tests
+main = defaultMainWithIohkTestMode tests
```
[the remark this answers is not in the document]
Comment (yogeshsajanikar) [reply ouroboros-consensus-byron-test/test/Main.hs:12]: `defaultMainWithTestMode` is required as it adds the option as an ingredient. Is it okay if the option is named 'test-mode'?

File: ouroboros-consensus-cardano-test/test/Test/ThreadNet/AllegraMary.hs [hunk 5]
```diff
@@ -164,12 +165,18 @@ oneTenthTestCount (QuickCheckTests n) = QuickCheckTests $
...
-      (if enabled then id else adjustOption oneTenthTestCount) $
+      askIohkTestMode $ flip adjustTestMode $
       testProperty name $ \setup ->
         prop_simple_allegraMary_convergence setup
     ]
 
+    where
+      adjustTestMode :: IohkTestMode -> TestTree -> TestTree
+      adjustTestMode = \case
+        Nightly -> resetQuickCheckTests id
```
[the remark this answers is not in the document]
Comment (yogeshsajanikar) [reply ouroboros-consensus-cardano-test/test/Test/ThreadNet/AllegraMary.hs:176]: I will add it to README.

File: ouroboros-consensus-shelley-test/test/Test/ThreadNet/Shelley.hs [hunk 6]
```diff
@@ -160,13 +162,13 @@ fifthTestCount (QuickCheckTests n) = QuickCheckTests $
...
 tests = testGroup "Shelley ThreadNet"
     [ let name = "simple convergence" in
-      askIohkNightlyEnabled $ \enabled ->
-      if enabled
-      then testProperty name $ \(NightlyTestSetup setup) ->
-             prop_simple_real_tpraos_convergence setup
-      else adjustOption fifthTestCount $
-           testProperty name $ \setup ->
-             prop_simple_real_tpraos_convergence setup
+      askIohkTestMode $ \case
```
[the remark this answers is not in the document]
Comment (yogeshsajanikar) [reply ouroboros-consensus-shelley-test/test/Test/ThreadNet/Shelley.hs:165]: Here the `NightlyTestSetup` is different for `Nightly` mode than other two. Hence, this deviation. Hope it's okay.

File: ouroboros-consensus-test/src/Test/Util/Nightly.hs (hunk 2 above)
Comment (yogeshsajanikar) [reply ouroboros-consensus-test/src/Test/Util/Nightly.hs:37]: Yes, I will remove. Thanks.

File: ouroboros-consensus-test/src/Test/Util/Nightly.hs [hunk 7]
```diff
@@ -35,7 +35,8 @@ instance IsOption IohkNightlyEnabled where
...
-  optionHelp = pure "Enable more expensive tests (specific to IOHK)"
+  optionHelp = pure "DEPRECATED Enable more expensive tests (specific to IOHK)\n\
```
[the remark this answers is not in the document]
Comment (yogeshsajanikar) [reply ouroboros-consensus-test/src/Test/Util/Nightly.hs:38]: Yes, intention was to link this option to `test-mode`. I did not want to break anything that's working now. But seeing that the option is not used anywhere else, I'd remove it alright.

File: ouroboros-consensus-test/src/Test/Util/TestMode.hs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+safeReadTestMode "dev"     = Just Dev
+safeReadTestMode _         = Nothing
+
+instance IsOption IohkTestMode where
+  defaultValue = Dev
+  parseValue = safeReadTestMode
+  optionName = pure "iohk-test-mode"
+  optionHelp = pure "Enable a test mode (specific to IOHK). \
+      \ The 'dev' mode runs default number of quickcheck tests (100), \
+      \ 'nightly' mode runs 1_000_000 quickcheck tests, and \
```
[the remark this answers is not in the document]
Comment (yogeshsajanikar) [reply ouroboros-consensus-test/src/Test/Util/TestMode.hs:75]: I agree. I tried to experiment with number of tests vs time. But each test needs different time settings, hence different number of tests. As a separate `tasty` library, I was thinking of a configuration file being supplied to `main` which would set parameters in a configuration file. We need some discussion on these numbers.

File: ouroboros-consensus-cardano-test/test/Test/ThreadNet/AllegraMary.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (yogeshsajanikar) [reply ouroboros-consensus-cardano-test/test/Test/ThreadNet/AllegraMary.hs]: Added a configuration where the tests are mentioned. So probably no need to add to README?

**Comment by yogeshsajanikar** [comment]
closes #3892

File: ouroboros-consensus-test/src/Test/Util/TestEnv.hs [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+instance IsOption TestEnv where
+  defaultValue = Dev
+  parseValue = safeReadTestMode
+  optionName = pure "test-env"
+  optionHelp = pure "Enable a test mode. \
+      \ The 'dev' mode runs default number of quickcheck tests (100), \
+      \ 'nightly' mode runs 100_000 quickcheck tests, and \
+      \ 'ci' mode runs 10_000 quickcheck tests. \
+      \ Individual tests can be adjusted to run less/more depending \
+      \ upon time taken to run them"
```
Comment (Jasagredo) [inline ouroboros-consensus-test/src/Test/Util/TestEnv.hs:67]:
```suggestion
      \ The 'dev' mode sets the default number of quickcheck tests to 100, \
      \ 'nightly' mode sets it to 100_000, and \
      \ 'ci' mode sets it to 10_000. \
      \ Individual tests are adjusted to run a number of tests proportional to the value above depending \
      \ on the time it takes to run them."
```

File: ouroboros-consensus-test/src/Test/Util/TestEnv.hs [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+withTestEnv TestEnvConfig{..} testTree = askOption $ \case
+      Nightly -> localOption (QuickCheckTests nightly) testTree
+      CI      -> localOption (QuickCheckTests ci) testTree
+      Dev     -> localOption (QuickCheckTests dev) testTree
+
+
+-- | This ingredient merely adds the 'TestEnv' 'Option' to the
+-- @tasty@ command-line parser.
+testEnvIngredient :: Ingredient
+testEnvIngredient = includingOptions [Option (Proxy :: Proxy TestEnv)]
```
Comment (Jasagredo) [inline ouroboros-consensus-test/src/Test/Util/TestEnv.hs:36]: Can't this be a definition local to `defaultMainWithTestEnv`

File: ouroboros-consensus-test/src/Test/Util/TestEnv.hs [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+import           Test.Tasty
+import           Test.Tasty.Ingredients
+import           Test.Tasty.Options
+import           Test.Tasty.QuickCheck
+
+-- | 'defaultMain' extended with 'iohkTestEnvIngredient'
+defaultMainWithTestEnv :: TestEnvConfig -> TestTree -> IO ()
+defaultMainWithTestEnv testConfig testTree =
+    defaultMainWithIngredients (testEnvIngredient : defaultIngredients) $ withTestEnv testConfig testTree
+
```
Comment (Jasagredo) [inline ouroboros-consensus-test/src/Test/Util/TestEnv.hs:23]: Nitpicking: we don't usually leave 2 blank lines between definitions

**Review by Jasagredo — APPROVED** [review]
I think this is fine (with a couple of minor comments). It would still be necessary to rename the PR and properly fill in the body of the PR description

**Pull request merged**